### ISSUE TITLE
New backend to better support rclone via ssh

### ISSUE THREAD

**Issue by allisonkarlitskaya** [restic/restic#4468]
Output of `restic version`
--------------------------
```
restic 0.15.2 compiled with go1.21rc3 on linux/amd64
```

What should restic do differently? Which functionality do you think we should add?
----------------------------------------------------------------------------------
The ability to use `rclone restic serve --stdio` via `ssh` by overriding the rclone command is really helpful, but it's a bit clunky.  In particular, the way that rclone takes `backend:path` as a single argument while `ssh` takes `host` separately is an impedance mismatch that requires adding extra parameters to the `ssh` command and ignoring the rclone backend name (which must be specified anyway).

It would be cool if there were a new (utterly trivial) dedicated backend which allowed you to avoid writing this:

```
restic -o rclone.program="ssh user@remotehost rclone serve restic --stdio" -r rclone:ignored:foo/bar
```

and instead, you could write this:

```
restic -r rclone-ssh:user@remotehost:foo/bar
```

Where `user@remotehost` gets passed directly through as the first parameter to `ssh` and the normal `rclone.program` and `rclone.args` get appended after.  There could even be a new `ssh.command` option or so, defaulting to `ssh` (searched in the `$PATH`).

So put another way:

```
restic -r rclone-ssh:HOST:PATH
```
would effectively turn into something like:

```
restic -r rclone:empty:%(PATH) -o rclone.command="%(ssh.command) %(HOST) %(rclone.command)" 
```

(plus or minus fussing around to make it actually work).

What are you trying to do? What problem would this solve?
---------------------------------------------------------

I'm doing restic backups via rclone on an ssh server, which is a reasonably common setup, I think.  My hosting provider even explicitly supports this configuration by allowing `rclone serve restic --stdio` as one of the few commands they support over SSH.

Did restic help you today? Did it make you happy in any way?
------------------------------------------------------------

restic is super awesome.  Even without the requested feature above (which basically just amounts to a bit of syntactic sugar) it's doing everything I need.  I've been looking for a long time to find a solution that I'd be comfortable with using for automated backups to the cloud with a reasonably design (de-duplication is super important for me, and solid encryption as well) and restic is the first solution that I trust the design of.

**Comment by arisudesu** [comment]
I think a better idea would be to leave the Rclone backend alone and instead add SSH support to the REST backend.

See, the Rclone backend was actually added for one specific purpose: running the Rclone process and communicating with it via REST over stdio. The ssh command substitution hack is quite clever, but not very user-friendly due to the need to enter additional flags. And in fact, it turns the original idea of the Rclone backend into a regular network connection to a REST server. What else can connect over the network using the REST protocol? Well, the REST backend itself.

Yes, you are right that you could add a separate backend for the rclone-ssh connection, but in reality it would be more like rest-ssh. And the `rest-ssh:HOST:PATH` scheme you propose is almost similar to the existing `rest:http(s)://host/path` scheme, except that the https protocol is changed to the SSH protocol. And thus, you can expand the scheme that the REST backend accepts to two formats - `rest:https://` and `rest:ssh://`. It would also be very similar to how we access Git repositories, which in turn is quite convenient and familiar, and has the same implementation details - connect to the server using the HTTPS or SSH transport protocol, then communicate using the Git protocol. Replace Git with REST and we get the same semantics.

In addition to this, I also want to point out an additional problem that I think could be fixed by introducing the SSH protocol into REST, which is the inability to use a subdirectory in Rclone via ssh command substitution. Look, the Rclone backend of the form `rclone:remote_name:path` assumes that path is passed to the rclone command in the arguments to remote, but this cannot be passed via ssh if the process `rclone serve restic --stdio` is started on the opposite side in some directory, in which case we can only access the repository in the root of the directory served by rclone. While in the REST backend there is no restriction on accessing subdirectories over HTTPS, and with the addition of SSH we could use multiple subdirectories this way (similar to Git repositories): `restic -r rest:ssh://rest@host :path1`, `restic -r rest:ssh://rest@host:path2` and so on. And for example, for different ssh keys we run `rclone serve restic --stdio` in different root directories, and get backup hosting for different users, with multiple repositories for each user. This is what I would like to do myself, but restic needs some improvements to access subdirectories correctly in Rclone over ssh.

tl;dr I propose to improve the REST backend by adding a second type of supported URL: `rest:ssh://rest@host:path` for connecting over ssh via the REST protocol, and at the same time be able to access subdirectories via ssh (we will get this without additional steps, it is already implemented in the REST protocol).

**Comment by MichaelEischer** [comment]
> I propose to improve the REST backend by adding a second type of supported URL: `rest:ssh://rest@host:path` for connecting over ssh via the REST protocol

It's not really connecting to rclone over ssh, but rather both starting rclone and directly connecting to it, or? (at least, unless a forced command is configured). That would somewhat suggest to use `rclone+ssh` as protocol scheme. Without a forced command there's currently also no way to specify options for rclone, although that would be rather simple to change.

I'm wondering whether it would be possible to add the functionality to the rclone backend. After all, that's what one would expect to start rclone. Conceptually something like `restic -o rclone.ssh=true -r rclone:user@host:remote:path` would also work. Although I have to admit that's not as neat as a custom protocol scheme.

What complicates things a bit is that we have two degrees of freedom here:
- the remote passed to rclone (ignored when using an ssh forced command)
- the path used by the REST backend

Using the `rest:ssh://[...]` approach that would result in something like `restic -r rest:ssh+rclone://user@host/path/in/repo -o rest.ssh-command="rclone serve restic --stdio remote"` (or maybe `rest.ssh-args="remote"` together with a default command).

When adding both options into the rclone backend, we'd end up with something like the following: `restic -r rclone:remote -o rclone.ssh="user@host" -o rclone.subpath=path/in/repo` which really looks weird. I'm not sure whether there's a nicer alternative when starting from the rclone backend.

**Comment by arisudesu** [comment]
I think I should've created a separate issue instead of posting in the comments. Then everything wouldn't be mixed up.

The author suggests simplifying the work with `rclone:` if he specifies the command to be launched via SSH, and he has the freedom to specify this command, i.e. the SSH server does not enforce a command bound to an SSH key on the server.

I, in turn, propose to add a connection method to `rest:` that will work over SSH, while the SSH server itself chooses which command to run for the connected client, maybe rclone with the given options, maybe something else, as long as it communicates using the Rest-server protocol. For example, when you use Git and SSH, the server runs git-shell for you; in this case, it would be `rclone serve restic --stdio` or the (fictional) `rest-shell`.

It could be implemented separately from rclone+ssh. The only thing that `rclone:` has in common with this is that from Restic we launch SSH and through it some remote process that communicates via the Rest-server protocol, and at the moment this can only be implemented through `rclone:` and the only possible command is `rclone serve command restic --stdio`, but more can be implemented in future.

Now, to the examples.

My current setup is the following:

  * On the machines from which I make backups, I run Restic this way:

        restic -o rclone.program="ssh" -o rclone.args="-i /path/to/this_machine_pkey rest@192.168.1.2" -r rclone:. backup /local/dir

    It connects the local Restic to a command that runs on 192.168.1.2 for a given key for **shared user named rest**.

  * On the machine that receives backups (192.168.1.2), there is a user **rest**, who has several keys entered into the ~/.ssh/authorized_keys file:

        command="/usr/bin/rclone serve restic --append-only --stdio /home/rest/machine1" ssh-rsa MACHINE_1_PUBKEY
        command="/usr/bin/rclone serve restic --append-only --stdio /home/rest/machine2" ssh-rsa MACHINE_2_PUBKEY
        ...

    It serves different repositories for different keys of this user. Backups of the machine with the MACHINE_1_PUBKEY key end up in the /home/rest/machine1 repository, backups of the MACHINE_2_PUBKEY machine in /home/rest/machine2, etc.

But in fact, I would like the server to not have to specify the path /home/rest/machine{1,2,...} for the keys, but to be able to specify one general path `/home/rest` for all keys, and specify the path on the machines themselves in the command, like: `restic -r rest:rest@192.168.1.2:machine1`, `restic -r rest:rest@192.168.1.2:machine2`, or even the absolute path: `restic -r rest:rest@192.168.0.1:/mnt/backups/myrepo`. We determine that we need to connect via SSH by the absence of "https:" in the repository parameter, and we use part of the parameter ":/mnt/backups/myrepo" or ":machine1" to determine which path to use in requests inside Rest-server protocol. At the same time, how to process this path in the requests is decided by the program that runs on the backup server for the keys. For example, if this is rclone, then it uses them as a path inside the remote (cloud), or rest-shell (fictional), which uses the path as the location of files in the filesystem on the machine that stores backups.

In the end the setup would look like this:

  * Backed up machines:

        restic -r rest:rest@192.168.1.2:machine1" backup /local/dir

  * Machine storing backups:

        command="/usr/bin/rclone serve restic --append-only --stdio" ssh-rsa MACHINE_1_PUBKEY
        command="/usr/bin/rclone serve restic --append-only --stdio" ssh-rsa MACHINE_2_PUBKEY
        ...

Which would look simpler than initial setup and looks alike other SSH protocols, e.g. git-shell, hg-ssh, svnserve etc.

We need to introduce the ssh.program parameter though, in the case a user needs to customize `ssh` command parameters, i.e. path to key or add some silly options for ssh like `-o StrictHostKeyChecking=no`; and ssh.remoteprogram allow user to specify command to run on the server, which may be combined like this:

    From:
        restic -r rest:SSHHOST:PATH
    To:
        ssh SSHHOST
    (SSH server must serve rest protocol to client)

    From:
        restic -r rest:SSHHOST:PATH -o ssh.program='ssh -o StrictHostKeyChecking=no'
    To:
        ssh -o StrictHostKeyChecking=no SSHHOST
    (SSH server must serve rest protocol to client)

    From:
        restic -r rest:SSHHOST:PATH -o ssh.remoteprogram='rclone serve restic --stdio remote:'
    To:
        ssh SSHHOST rclone serve restic --stdio remote:
    (clients runs specific command over ssh that must serve rest protocol, which server may or may not allow to run)

Thus either enforcing backing up over the connection (a "SSH backup server" solution) or allowing user to serve remote directory to local restic (a personal SSH account solution which may serve like a file storage).
